File: pipeline/checker.py
"""CF-style token checker (mirrors the judge contract in docs/JUDGE_API.md)."""


def tokens(text: str):
    return text.split()


def check(expected: str, actual: str, float_eps=None) -> bool:
    et, at = tokens(expected), tokens(actual)
    if len(et) != len(at):
        return False
    for e, a in zip(et, at):
        if e == a:
            continue
        if e.lower() in ("yes", "no") and e.lower() == a.lower():
            continue
        if float_eps is not None:
            try:
                fe, fa = float(e), float(a)
                if abs(fe - fa) <= float_eps or (
                    fe != 0 and abs(fe - fa) / abs(fe) <= float_eps
                ):
                    continue
            except ValueError:
                pass
        return False
    return True

File: judge/src/checker.ts
export interface CheckResult {
  ok: boolean;
  note?: string;
}

const YESNO = new Set(["yes", "no"]);

function isNumeric(tok: string): boolean {
  if (tok.length === 0) return false;
  const n = Number(tok);
  return !Number.isNaN(n) && /^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/.test(tok);
}

/**
 * CF-style token comparison. Splits both outputs on whitespace (so trailing
 * whitespace/newlines are ignored), compares token by token. YES/NO tokens are
 * case-insensitive. If floatEps is set, numeric tokens compare with abs/rel eps.
 */
export function check(
  expected: string,
  actual: string,
  floatEps?: number | null
): CheckResult {
  const exp = expected.split(/\s+/).filter((t) => t.length > 0);
  const act = actual.split(/\s+/).filter((t) => t.length > 0);
  if (exp.length !== act.length) {
    return {
      ok: false,
      note: `expected ${exp.length} tokens, got ${act.length}`,
    };
  }
  for (let i = 0; i < exp.length; i++) {
    const e = exp[i];
    const a = act[i];
    if (e === a) continue;
    const el = e.toLowerCase();
    const al = a.toLowerCase();
    if (YESNO.has(el) && el === al) continue;
    if (floatEps != null && isNumeric(e) && isNumeric(a)) {
      const en = Number(e);
      const an = Number(a);
      const diff = Math.abs(en - an);
      if (diff <= floatEps || diff <= floatEps * Math.abs(en)) continue;
      return {
        ok: false,
        note: `token ${i + 1} differs: expected ${e}, got ${a} (eps ${floatEps})`,
      };
    }
    return { ok: false, note: `token ${i + 1} differs: expected ${e}, got ${a}` };
  }
  return { ok: true };
}
